common/namespace/registry.go

```diff
@@ -483,6 +483,200 @@ func (r *registry) getNamespaceByIDLocked(id ID) (*Namespace, error) {
+	ns, err := r.getNamespaceByNamePersistence(name)
+	if err != nil {
+		return nil, err
+	}
+
+	// update main caches if found
+	r.updateCachesSingleNamespace(ns)
+
+	return ns, nil
+}
+
+func (r *registry) checkCachesByName(name Name) (*Namespace, error) {
+	// check main caches
+	cacheHit, cacheErr := r.getNamespace(name)
+	if cacheErr == nil {
+		return cacheHit, nil
+	}
+
+	// check readthrough cache to prevent excessive requests to persistence layer
+	prevErr := r.getPreviousReadthroughError(name.String())
+	if prevErr != nil {
+		return nil, prevErr
+	}
+
+	// missed both caches, should readthrough
+	return nil, nil
+}
+
+// getOrPutNamespaceByID retrieves the information from the cache if it exists or reads through
+// to the persistence layer and updates caches if it doesn't
+func (r *registry) getOrPutNamespaceByID(id ID) (*Namespace, error) {
+	// check caches
+	cachedNS, cachedErr := r.checkCachesByID(id)
+	if cachedNS != nil || cachedErr != nil {
+		return cachedNS, cachedErr
+	}
+
+	r.readthroughRequestLock.Lock()
+	defer r.readthroughRequestLock.Unlock()
+
+	// check caches again in case there was an update while waiting
+	cachedNS, cachedErr = r.checkCachesByID(id)
+	if cachedNS != nil || cachedErr != nil {
+		return cachedNS, cachedErr
+	}
+
+	// readthrough to persistence layer and update readthrough cache if not found
+	ns, err := r.getNamespaceByIDPersistence(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// update main caches if found
+	r.updateCachesSingleNamespace(ns)
+
+	return ns, nil
+}
+
+func (r *registry) checkCachesByID(id ID) (*Namespace, error) {
+	// check main caches
+	cacheHit, cacheErr := r.getNamespaceByID(id)
+	if cacheErr == nil {
+		return cacheHit, nil
+	}
+
+	// check readthrough error cache to prevent excessive requests to persistence layer
+	prevErr := r.getPreviousReadthroughError(id.String())
+	if prevErr != nil {
+		return nil, prevErr
+	}
+
+	// missed both caches, should readthrough
+	return nil, nil
+}
+
+func (r *registry) getPreviousReadthroughError(handle string) error {
+	r.readthroughCacheLock.RLock()
+	defer r.readthroughCacheLock.RUnlock()
+
+	if prevErr, ok := r.readthroughErrorsCache.Get(handle).(error); ok {
+		return prevErr
+	}
+	return nil
+}
+
+func (r *registry) updateCachesSingleNamespace(ns *Namespace) {
+	var stateChangeCallbacks []StateChangeCallbackFn
+
+	r.cacheLock.Lock()
+	if curEntry, ok := r.cacheByID.Get(ns.ID()).(*Namespace); ok {
+		if curEntry.ConfigVersion() >= ns.ConfigVersion() {
+			// More up to date version already put in cache by refresh
+			r.cacheLock.Unlock()
+			return
+		}
+	}
+
+	oldNS := r.updateIDToNamespaceCache(r.cacheByID, ns.ID(), ns)
+	r.cacheNameToID.Put(ns.Name(), ns.ID())
+	if namespaceStateChanged(oldNS, ns) {
+		stateChangeCallbacks = mapAnyValues(r.stateChangeCallbacks)
+	}
+	r.cacheLock.Unlock()
+
+	// call state change callbacks
+	for _, cb := range stateChangeCallbacks {
+		cb(ns, false)
```

Comment:
- I don't think we should run callbacks synchronously in the GetNamespace call.
- This implementation requires callbacks to support concurrency calls, and previously we don't have this requirement I think? cc @dnr

Reply:
Should we put them behind `cacheLock`? We must run the callbacks when updating the cache or they may get missed by the refresh loop

Reply:
Concurrent calls: There are three callbacks so far. Two clearly have no concurrency issues (sending on a channel, starting a goroutine). The third (in history engine) seems okay too (shard context has its own lock, queue processor rescheduler has its own lock) but I'm not 100% sure. In general it seems okay to add this requirement even if we always call from refresh loop.

@pdoerner What do you mean they may get missed by the refresh loop? The issue I see is that the refresh loop might run before this goroutine gets around to calling them. But since we updated the cache here, I think that wouldn't do anything bad? It will not call them, and then we'll do it here?

Overall the intention was that the semantics of this callback should be asynchronous and relaxed (compared to the old one). So I agree that it's probably bad to call them synchronously. So maybe we could just put this little bit in a goroutine? I don't really like that either.

We could mark/store these to be called by the refresh loop next time, and then trigger the refresh loop? A little more complicated but safer?

Reply:
If we update the cache during readthrough, without running callbacks, and there are no updates before the next refresh loop, then the next refresh will not see any change between the namespace it got from `ListNamespaces` and what was already in the cache, so it will not run the callbacks. So I think we need some mechanism to run them whenever we make a change to the cache. Why don't you like putting this bit in a separate goroutine?